From 4a003f85d92b1e590c1ae48402304c27ad0fe2d4 Mon Sep 17 00:00:00 2001
From: Daniil Klimuk <daniil.klimuk@3mdeb.com>
Date: Mon, 30 Dec 2024 11:56:00 +0100
Subject: [PATCH] Patch for RK3566 on version 3.13.0

This patch cannot be considered as complete because it does not have
right MMU coonfiguration. Therefore it is not recommended to use it
to produce OPTEE OS image.

On the other hand, building optee_os repository artifacts produces not
only OPTEE OS image but TA Development Kit and TAs as well. The TA
Development Kit is being used in optee_test artifacts compilation as
well.These artifacts does not use MMU (because configuration of MMU is
the OPTEE OS responsibility), but might utilize some other OPTEE OS
configuration (e.g. UART configuration).

Upstream-Status: Inappropriate other

This is a local fix, no need to upstream.

Signed-off-by: Daniil Klimuk <daniil.klimuk@3mdeb.com>
---
 core/arch/arm/plat-rockchip/conf.mk           | 23 ++++++++++
 core/arch/arm/plat-rockchip/platform_config.h | 28 ++++++++++++
 core/arch/arm/plat-rockchip/platform_rk3566.c | 44 +++++++++++++++++++
 core/arch/arm/plat-rockchip/sub.mk            |  1 +
 4 files changed, 96 insertions(+)
 create mode 100644 core/arch/arm/plat-rockchip/platform_rk3566.c

diff --git a/core/arch/arm/plat-rockchip/conf.mk b/core/arch/arm/plat-rockchip/conf.mk
index f180d63acdf4..4692b41b2966 100644
--- a/core/arch/arm/plat-rockchip/conf.mk
+++ b/core/arch/arm/plat-rockchip/conf.mk
@@ -55,6 +55,29 @@ CFG_SHMEM_SIZE   ?= 0x00400000
 CFG_EARLY_CONSOLE ?= n
 endif

+ifeq ($(PLATFORM_FLAVOR),rk3566)
+include core/arch/arm/cpu/cortex-armv8-0.mk
+arm32-platform-cpuarch  := cortex-a55
+
+$(call force,CFG_TEE_CORE_NB_CORE,4)
+# cortex-a55 supports GICv3 or GICv4. There is no support for GICv4 in OP-TEE,
+# so use GICv3:
+$(call force,CFG_ARM_GICV3,y)
+
+CFG_TZDRAM_START ?= 0x08400000
+CFG_TZDRAM_SIZE  ?= 0x00e00000
+CFG_SHMEM_START  ?= 0x09200000
+CFG_SHMEM_SIZE   ?= 0x00200000
+
+CFG_EARLY_CONSOLE ?= y
+CFG_EARLY_CONSOLE_BASE ?= UART2_BASE
+CFG_EARLY_CONSOLE_SIZE ?= UART2_SIZE
+CFG_EARLY_CONSOLE_BAUDRATE ?= 1500000
+CFG_EARLY_CONSOLE_CLK_IN_HZ ?= 24000000
+
+CFG_WITH_ARM_TRUSTED_FW ?= y
+endif
+
 ifeq ($(platform-flavor-armv8),1)
 $(call force,CFG_ARM64_core,y)
 $(call force,CFG_WITH_ARM_TRUSTED_FW,y)
diff --git a/core/arch/arm/plat-rockchip/platform_config.h b/core/arch/arm/plat-rockchip/platform_config.h
index 88a53b40740c..6cb91bd64814 100644
--- a/core/arch/arm/plat-rockchip/platform_config.h
+++ b/core/arch/arm/plat-rockchip/platform_config.h
@@ -84,6 +84,34 @@
 #define FIREWALL_DDR_BASE	0xff534000
 #define FIREWALL_DDR_SIZE	SIZE_K(16)

+#elif defined(PLATFORM_FLAVOR_rk3566)
+
+/* This information is from Technical Reference Manual for RK3568, which is said
+ * to be very close in configuration to RK3566. RTM for RK3566 has net been
+ * found. */
+
+#define MMIO_BASE		0xF0000000
+
+#define UART0_BASE		(MMIO_BASE + 0x0DD50000)
+#define UART0_SIZE		SIZE_K(64)
+
+#define UART1_BASE		(MMIO_BASE + 0x0E650000)
+#define UART1_SIZE		SIZE_K(64)
+
+#define UART2_BASE		(MMIO_BASE + 0x0E660000)
+#define UART2_SIZE		SIZE_K(64)
+
+#define UART3_BASE		(MMIO_BASE + 0x0E670000)
+#define UART3_SIZE		SIZE_K(64)
+
+#define SGRF_BASE		(MMIO_BASE + 0x0DC60000)
+#define SGRF_SIZE		SIZE_K(64)
+
+#define GIC_BASE		(MMIO_BASE + 0x0d400000)
+#define GIC_SIZE		SIZE_M(4)
+#define GICD_BASE		(GIC_BASE)
+#define GICR_BASE		(GIC_BASE + 0x60000)
+
 #else
 #error "Unknown platform flavor"
 #endif
diff --git a/core/arch/arm/plat-rockchip/platform_rk3566.c b/core/arch/arm/plat-rockchip/platform_rk3566.c
new file mode 100644
index 000000000000..d403ec7db018
--- /dev/null
+++ b/core/arch/arm/plat-rockchip/platform_rk3566.c
@@ -0,0 +1,44 @@
+#include <common.h>
+#include <io.h>
+#include <kernel/panic.h>
+#include <mm/core_memprot.h>
+#include <platform.h>
+#include <platform_config.h>
+
+#define SGRF_DDRRGN_CON0_16(n)		((n) * 4)
+#define SGRF_DDR_RGN_0_16_WMSK		GENMASK_32(11, 0)
+
+register_phys_mem_pgdir(MEM_AREA_IO_SEC, SGRF_BASE, SGRF_SIZE);
+
+int platform_secure_ddr_region(int rgn, paddr_t st, size_t sz)
+{
+	vaddr_t sgrf_base = (vaddr_t)phys_to_virt_io(SGRF_BASE);
+	paddr_t ed = st + sz;
+	uint32_t st_mb = st / SIZE_M(1);
+	uint32_t ed_mb = ed / SIZE_M(1);
+
+	if (!sgrf_base)
+		panic();
+
+	assert(rgn <= 7);
+	assert(st < ed);
+
+	/* Check aligned 1MB */
+	assert(st % SIZE_M(1) == 0);
+	assert(ed % SIZE_M(1) == 0);
+
+	DMSG("protecting region %d: 0x%lx-0x%lx", rgn, st, ed);
+
+	/* Set ddr region addr start */
+	io_write32(sgrf_base + SGRF_DDRRGN_CON0_16(rgn),
+		   BITS_WITH_WMASK(st_mb, SGRF_DDR_RGN_0_16_WMSK, 0));
+
+	/* Set ddr region addr end */
+	io_write32(sgrf_base + SGRF_DDRRGN_CON0_16(rgn + 8),
+		   BITS_WITH_WMASK((ed_mb - 1), SGRF_DDR_RGN_0_16_WMSK, 0));
+
+	io_write32(sgrf_base + SGRF_DDRRGN_CON0_16(16),
+		   BIT_WITH_WMSK(rgn));
+
+	return 0;
+}
diff --git a/core/arch/arm/plat-rockchip/sub.mk b/core/arch/arm/plat-rockchip/sub.mk
index 74f8cf5a920c..ac46347976dd 100644
--- a/core/arch/arm/plat-rockchip/sub.mk
+++ b/core/arch/arm/plat-rockchip/sub.mk
@@ -4,6 +4,7 @@ srcs-y += platform.c
 srcs-$(PLATFORM_FLAVOR_px30) += platform_px30.c
 srcs-$(PLATFORM_FLAVOR_rk322x) += platform_rk322x.c
 srcs-$(PLATFORM_FLAVOR_rk3399) += platform_rk3399.c
+srcs-$(PLATFORM_FLAVOR_rk3566) += platform_rk3566.c

 ifeq ($(PLATFORM_FLAVOR),rk322x)
 srcs-y += plat_init.S
--
2.47.1
